File: 442*find-all-duplicates-in-an-array/solution.py
class Solution(object):
    def findDuplicates(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        ret = []
        
        for idx, n in enumerate(nums):
            idx = abs(n) - 1
            if nums[idx] > 0:
                nums[idx] *= -1
            else:
                ret.append(abs(n))
                
        return ret

File: 442*find-all-duplicates-in-an-array/solution.java

class Solution {
    public List<Integer> findDuplicates(int[] nums) {
        List<Integer> ret = new ArrayList<Integer>();
        
        for (int i=0; i < nums.length; i++) {
            int idx = Math.abs(nums[i]) -1;          
            if (nums[idx] < 0) {
                ret.add(idx+1);
            }
            nums[idx] *= -1;
        }
        
        return ret;
    }
}
